File: python/aethernet/bittorrent/krpc.py
# SPDX-License-Identifier: MIT
"""KRPC (BEP-5) DHT messages over bencode."""
from __future__ import annotations

from . import bencode

QUERY = "q"
RESPONSE = "r"
ERROR = "e"


def encode_query(tx: bytes, method, args: dict) -> bytes:
    m = method.encode("utf-8") if isinstance(method, str) else method
    return bencode.encode({b"t": tx, b"y": b"q", b"q": m, b"a": args})


def encode_response(tx: bytes, response: dict) -> bytes:
    return bencode.encode({b"t": tx, b"y": b"r", b"r": response})


def encode_error(tx: bytes, code: int, message) -> bytes:
    msg = message.encode("utf-8") if isinstance(message, str) else message
    return bencode.encode({b"t": tx, b"y": b"e", b"e": [code, msg]})


def decode(data: bytes) -> dict:
    d = bencode.decode(data)
    if not isinstance(d, dict):
        raise ValueError("KRPC message is not a dictionary")
    y = d.get(b"y")
    out = {"transaction_id": d.get(b"t"), "type": y.decode("utf-8") if isinstance(y, bytes) else None}
    if y == b"q":
        out["method"] = d.get(b"q", b"").decode("utf-8")
        out["arguments"] = d.get(b"a", {})
    elif y == b"r":
        out["response"] = d.get(b"r", {})
    elif y == b"e":
        e = d.get(b"e", [])
        if len(e) >= 2:
            out["error_code"] = e[0]
            out["error_message"] = e[1].decode("utf-8") if isinstance(e[1], bytes) else e[1]
    return out

File: kotlin/src/main/kotlin/aethernet/bittorrent/Krpc.kt
// SPDX-License-Identifier: MIT

package aethernet.bittorrent

/** The KRPC message kind (the "y" field). */
enum class KrpcType { QUERY, RESPONSE, ERROR }

/**
 * A KRPC message (BEP-5): a bencoded dict with "t" (transaction id), "y" (q/r/e), and
 * one of a query ("q"+"a"), response ("r"), or error ("e"). The Kotlin port of
 * `go/bittorrent/krpc.go`.
 */
class KrpcMessage(
    val transactionId: ByteArray,
    val type: KrpcType,
    val method: String? = null,       // query: the "q" method name
    val arguments: BDict? = null,     // query: "a"
    val response: BDict? = null,      // response: "r"
    val errorCode: Long = 0,          // error: e[0]
    val errorMessage: String? = null, // error: e[1]
) {
    /** Serializes the message to canonical bencode. */
    fun encode(): ByteArray {
        val d = BDict()
        d.add("t", BStr(transactionId))
        when (type) {
            KrpcType.QUERY -> {
                d.add("y", BStr("q"))
                d.add("q", BStr(method ?: ""))
                d.add("a", arguments ?: BDict())
            }
            KrpcType.RESPONSE -> {
                d.add("y", BStr("r"))
                d.add("r", response ?: BDict())
            }
            KrpcType.ERROR -> {
                d.add("y", BStr("e"))
                d.add("e", BList(listOf(BInt(errorCode), BStr(errorMessage ?: ""))))
            }
        }
        return Bencode.encode(d)
    }

    companion object {
        /** Parses a KRPC message. */
        fun decode(data: ByteArray): KrpcMessage {
            val d = Bencode.decode(data).asDict()

            val transactionId = (d.get("t") ?: throw BencodeException("KRPC message has no 't'")).asBytes()
            val y = (d.get("y") ?: throw BencodeException("KRPC message has no 'y'")).asText()

            return when (y) {
                "q" -> {
                    val method = (d.get("q") ?: throw BencodeException("KRPC query has no 'q'")).asText()
                    val arguments = (d.get("a") as? BDict)
                    KrpcMessage(transactionId, KrpcType.QUERY, method = method, arguments = arguments)
                }
                "r" -> KrpcMessage(transactionId, KrpcType.RESPONSE, response = d.get("r") as? BDict)
                "e" -> {
                    var code = 0L
                    var message = ""
                    (d.get("e") as? BList)?.let { list ->
                        if (list.items.size >= 2) {
                            code = list.items[0].asInt()
                            message = list.items[1].asText()
                        }
                    }
                    KrpcMessage(transactionId, KrpcType.ERROR, errorCode = code, errorMessage = message)
                }
                else -> throw BencodeException("unknown KRPC y=\"$y\"")
            }
        }
    }
}
